hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/InProcessHapiTestNode.java

```diff
@@ -90,21 +151,123 @@ public void waitForActive(long seconds) {
+            // way to shut down a node. Normally, nodes are shutdown by existing the JVM. However, we don't want to
+            // do that here, because we're in-process! So what I'm going to do is:
+            //  1. Search through all the threads in the JVM to find all of them with a callstack with "WorkerThread"
+            //     in the stack. This means they were called by the WorkerThread. Only objects that are part of the
+            //     node will be in this callstack.
+            //  2. For each such thread, stop it. Thread.stop is deprecated and discouraged, because it is almost
+            //     always the wrong thing. In our case though, NOTHING in the Junit JVM is working with any locks or
+            //     semaphores, etc. in the node. So we should be safe.
+            //  3. There are some places in the node software that uses statics -- like MerkleDb. I've got to try to
+            //     reset those.
+            //
+            // This is an error-prone approach, because at any time someone might add new static state to the node and
+            // fail to update this code accordingly. But it's the best we can do without removing all static state and
+            // having a clean shutdown procedure for the node. Fixing that should be a priority, allowing us to simplify
+            // this code and make it more foolproof.
+
+            //noinspection deprecation
+            final var threadsToStop = Thread.getAllStackTraces().entrySet().stream()
+                    .filter(entry -> {
+                        for (final var stackTraceElement : entry.getValue()) {
+                            final var className = stackTraceElement.getClassName();
+                            if (className.contains("WorkerThread") || className.contains("com.swirlds")) {
+                                return true;
+                            }
+                        }
+                        return false;
+                    })
+                    .map(Map.Entry::getKey)
+                    .toList();
+
+            threadsToStop.forEach(th -> {
+                if (th instanceof Stoppable s) {
+                    s.stop();
+                }
+            });
+            threadsToStop.forEach(Thread::interrupt);
+            threadsToStop.forEach(Thread::stop);
+
+            MerkleDb.setDefaultPath(null);
+            ConstructableRegistry.getInstance().reset();
+        }
+    }
+
+    @Override
+    public void waitForShutdown(long seconds) throws TimeoutException {
+        final var waitUntil = System.currentTimeMillis() + (seconds * 1000);
+        while (th != null && th.hedera != null && th.hedera.isActive()) {
+            if (System.currentTimeMillis() > waitUntil) {
+                throw new TimeoutException(
+                        "node " + nodeId + ": Waited " + seconds + " seconds, but node did not shut down!");
+            }
+
+            try {
```

Comment:
That would most likely be a condition variable.  It's more complicated than a loop, however, and might not be worth that complexity in test code.

An alternative would be for `Hedera.shutdown` to return a CompletableFuture and wait for that to complete (with timeout).  That might be a cleaner approach, as it could cleanly compose the different waits behind the server shutdown into a single item.